File: server/src/scripts/test-all-models.ts
/**
 * Probe every enabled model with a minimal request to find broken model IDs.
 * Usage: npx tsx src/scripts/test-all-models.ts
 */
import { initDb, getDb } from '../db/index.js';
import { decrypt } from '../lib/crypto.js';
import { getProvider } from '../providers/index.js';
import * as schema from '../db/schema.js';
import { eq, and, exists, asc } from 'drizzle-orm';
import type { Platform } from '@freellmapi/shared/types.js';

initDb();
const db = getDb();

const modelsWithKeys = db.select({
  id: schema.models.id,
  platform: schema.models.platform,
  modelId: schema.models.modelId,
  displayName: schema.models.displayName
})
.from(schema.models)
.where(and(
  eq(schema.models.enabled, 1),
  exists(
    db.select()
      .from(schema.apiKeys)
      .where(and(
        eq(schema.apiKeys.platform, schema.models.platform),
        eq(schema.apiKeys.enabled, 1)
      ))
  )
))
.orderBy(asc(schema.models.intelligenceRank), asc(schema.models.platform))
.all();

const results: { row: typeof modelsWithKeys[0]; ok: boolean; ms: number; error?: string; reply?: string }[] = [];

for (const row of modelsWithKeys) {
  const keyRow = db.select({
    encryptedKey: schema.apiKeys.encryptedKey,
    iv: schema.apiKeys.iv,
    authTag: schema.apiKeys.authTag
  })
  .from(schema.apiKeys)
  .where(and(
    eq(schema.apiKeys.platform, row.platform),
    eq(schema.apiKeys.enabled, 1)
  ))
  .orderBy(asc(schema.apiKeys.id))
  .limit(1)
  .get();

  if (!keyRow) {
    results.push({ row, ok: false, ms: 0, error: 'no key' });
    continue;
  }

  const apiKey = decrypt(keyRow.encryptedKey, keyRow.iv, keyRow.authTag);
  const provider = getProvider(row.platform as Platform);
  if (!provider) {
    results.push({ row, ok: false, ms: 0, error: 'no provider' });
    continue;
  }

  const start = Date.now();
  try {
    const res = await provider.chatCompletion(apiKey, [{ role: 'user', content: 'hi' }], row.modelId, { max_tokens: 5 });
    const replyContent = res.choices?.[0]?.message?.content;
    const reply = typeof replyContent === 'string' ? replyContent.slice(0, 40) : '';
    results.push({ row, ok: true, ms: Date.now() - start, reply });
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    results.push({ row, ok: false, ms: Date.now() - start, error: message.slice(0, 200) });
  }
}

console.log('\n=== Results ===\n');
const pad = (s: string, n: number) => s.length > n ? s.slice(0, n - 1) + '…' : s.padEnd(n);
for (const r of results) {
  const status = r.ok ? '✓' : '✗';
  console.log(`${status} ${pad(r.row.platform, 12)} ${pad(r.row.modelId, 52)} ${String(r.ms).padStart(5)}ms  ${r.ok ? `"${r.reply}"` : r.error}`);
}
const okCount = results.filter(r => r.ok).length;
console.log(`\n${okCount}/${results.length} models working\n`);

process.exit(0);

File: crates/server/src/bin/test_all_models.rs
//! Probe every enabled model that has an enabled API key.

use std::time::Instant;

use server::crypto;
use server::db;
use server::env;
use server::types::{ChatMessage, CompletionOptions};

#[derive(Debug, Clone)]
struct ModelTarget {
    platform: String,
    model_id: String,
}

#[derive(Debug)]
struct ProbeResult {
    target: ModelTarget,
    ok: bool,
    elapsed_ms: u128,
    detail: String,
}

#[tokio::main]
async fn main() {
    env::init();
    server::logger::init();

    db::init_db(None).expect("failed to initialize database");
    db::run_in_transaction(|conn| {
        db::run_migrations(conn);
        Ok::<(), rusqlite::Error>(())
    })
    .await
    .expect("failed to run database migrations");

    let targets = load_targets().await;
    let mut results = Vec::with_capacity(targets.len());

    for target in targets {
        results.push(probe_model(target).await);
    }

    println!("\n=== Results ===\n");
    for result in &results {
        let status = if result.ok { '✓' } else { '✗' };
        println!(
            "{status} {} {} {:>5}ms  {}",
            pad(status_text(&result.target.platform, 12), 12),
            pad(status_text(&result.target.model_id, 52), 52),
            result.elapsed_ms,
            result.detail
        );
    }

    let working = results.iter().filter(|result| result.ok).count();
    println!("\n{working}/{} models working\n", results.len());
}

async fn load_targets() -> Vec<ModelTarget> {
    let conn = db::db().lock().await;
    let mut statement = conn
        .prepare(
            "SELECT m.platform, m.model_id
             FROM models AS m
             WHERE m.enabled = 1
               AND EXISTS (
                   SELECT 1 FROM api_keys AS k
                   WHERE k.platform = m.platform AND k.enabled = 1
               )
             ORDER BY m.intelligence_rank ASC, m.platform ASC",
        )
        .expect("failed to query enabled models");

    statement
        .query_map([], |row| {
            Ok(ModelTarget {
                platform: row.get(0)?,
                model_id: row.get(1)?,
            })
        })
        .expect("failed to read enabled models")
        .collect::<rusqlite::Result<Vec<_>>>()
        .expect("failed to collect enabled models")
}

async fn load_key(target: &ModelTarget) -> Option<(String, String, String)> {
    let conn = db::db().lock().await;
    conn.query_row(
        "SELECT encrypted_key, iv, auth_tag
         FROM api_keys
         WHERE platform = ?1 AND enabled = 1
         ORDER BY id ASC
         LIMIT 1",
        rusqlite::params![target.platform],
        |row| Ok((row.get(0)?, row.get(1)?, row.get(2)?)),
    )
    .ok()
}

async fn probe_model(target: ModelTarget) -> ProbeResult {
    let Some((encrypted_key, iv, auth_tag)) = load_key(&target).await else {
        return ProbeResult {
            target,
            ok: false,
            elapsed_ms: 0,
            detail: "no key".to_string(),
        };
    };

    let api_key = match crypto::decrypt(&encrypted_key, &iv, &auth_tag) {
        Ok(key) => key,
        Err(error) => {
            return ProbeResult {
                target,
                ok: false,
                elapsed_ms: 0,
                detail: format!("decrypt failed: {error}"),
            };
        }
    };

    let Some(provider) = server::providers::get_provider(&target.platform).await else {
        return ProbeResult {
            target,
            ok: false,
            elapsed_ms: 0,
            detail: "no provider".to_string(),
        };
    };

    let started = Instant::now();
    let response = provider
        .chat_completion(
            &api_key,
            &[ChatMessage::text("user", "hi")],
            &target.model_id,
            &CompletionOptions {
                max_tokens: Some(5),
                ..CompletionOptions::default()
            },
        )
        .await;
    let elapsed_ms = started.elapsed().as_millis();

    match response {
        Ok(response) => {
            let reply = response
                .choices
                .first()
                .and_then(|choice| choice.message.content.as_text())
                .map(one_line)
                .unwrap_or_default();
            ProbeResult {
                target,
                ok: true,
                elapsed_ms,
                detail: format!("\"{}\"", truncate(&reply, 40)),
            }
        }
        Err(error) => ProbeResult {
            target,
            ok: false,
            elapsed_ms,
            detail: truncate(&one_line(&error.to_string()), 200),
        },
    }
}

fn status_text(value: &str, width: usize) -> String {
    truncate(value, width)
}

fn pad(value: String, width: usize) -> String {
    let len = value.chars().count();
    if len >= width {
        value
    } else {
        format!("{value}{}", " ".repeat(width - len))
    }
}

fn truncate(value: &str, max_chars: usize) -> String {
    let mut chars = value.chars();
    let truncated: String = chars.by_ref().take(max_chars).collect();
    if chars.next().is_some() {
        let mut shortened: String = truncated
            .chars()
            .take(max_chars.saturating_sub(1))
            .collect();
        shortened.push('…');
        shortened
    } else {
        truncated
    }
}

fn one_line(value: &str) -> String {
    value
        .chars()
        .map(|character| match character {
            '\n' | '\r' | '\t' => ' ',
            other => other,
        })
        .collect()
}
